From 1de86c08d63889bcf900cb2213e3c68e76ae7df8 Mon Sep 17 00:00:00 2001
From: Jason Rogena <j.rogena@cgiar.org>
Date: Thu, 3 Apr 2014 12:05:34 +0300
Subject: [PATCH 141/213] Version 0.8.16: implemented feature #150. Milk
 production history activity now has a back button, limits to data from past 7
 days orders the data using the date and shows a tables showing production
 totals for each cow

---
 MistroFarmer/src/main/AndroidManifest.xml          |   2 +-
 .../farmer/MilkProcutionHistoryActivity.java       | 201 ++++++++++++++++++++-
 .../layout/activity_milk_production_history.xml    | 153 +++++++++++-----
 MistroFarmer/src/main/res/values/strings_en.xml    |   1 +
 MistroFarmer/src/main/res/values/strings_sw.xml    |   1 +
 5 files changed, 309 insertions(+), 49 deletions(-)

diff --git a/MistroFarmer/src/main/AndroidManifest.xml b/MistroFarmer/src/main/AndroidManifest.xml
index 8ab3278..ef02a03 100644
--- a/MistroFarmer/src/main/AndroidManifest.xml
+++ b/MistroFarmer/src/main/AndroidManifest.xml
@@ -2,7 +2,7 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="org.cgiar.ilri.mistro.farmer"
     android:versionCode="2"
-    android:versionName="0.8.15" >
+    android:versionName="0.8.16" >
 
     <uses-sdk
         android:minSdkVersion="7"
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MilkProcutionHistoryActivity.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MilkProcutionHistoryActivity.java
index dea6196..09deaeb 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MilkProcutionHistoryActivity.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MilkProcutionHistoryActivity.java
@@ -10,6 +10,7 @@ import android.util.DisplayMetrics;
 import android.util.Log;
 import android.view.Gravity;
 import android.view.View;
+import android.widget.Button;
 import android.widget.TableLayout;
 import android.widget.TableRow;
 import android.widget.TextView;
@@ -32,20 +33,29 @@ import java.util.ArrayList;
 import java.util.Calendar;
 import java.util.Date;
 import java.util.GregorianCalendar;
+import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 
-public class MilkProcutionHistoryActivity extends SherlockActivity
+public class MilkProcutionHistoryActivity extends SherlockActivity implements View.OnClickListener
 {
     private static final String TAG="MIlkProductionHistoryActivity";
     private Menu menu;
     private TextView dateTV;
     private TextView cowNameTV;
     private TextView timeTV;
-    private  TextView quantityTV;
+    private TextView quantityTV;
     private List<String> productionHistoryIDs;
     private String noDataWarning;
     private DisplayMetrics metrics;
     private TableLayout productionHistoryTL;
+    private TableLayout productionTotalTL;
+    private TextView totalDateTV;
+    private TextView totalCowNameTV;
+    private TextView totalQuantityTV;
+    private Button backB;
+
     private String[] times;
     private String todayText;
     private String yesterdayText;
@@ -66,6 +76,12 @@ public class MilkProcutionHistoryActivity extends SherlockActivity
         timeTV=(TextView)findViewById(R.id.time_tv);
         quantityTV=(TextView)findViewById(R.id.quantity_tv);
         productionHistoryTL=(TableLayout)findViewById(R.id.production_history_tl);
+        productionTotalTL = (TableLayout)findViewById(R.id.production_total_tl);
+        totalDateTV = (TextView)findViewById(R.id.total_date_tv);
+        totalCowNameTV = (TextView)findViewById(R.id.total_cow_name_tv);
+        totalQuantityTV = (TextView)findViewById(R.id.total_quantity_tv);
+        backB = (Button)findViewById(R.id.back_b);
+        backB.setOnClickListener(this);
 
         initTextInViews();
         fetchProductionHistory();
@@ -106,6 +122,11 @@ public class MilkProcutionHistoryActivity extends SherlockActivity
         cowNameTV.setText(Locale.getStringInLocale("cow",this));
         timeTV.setText(Locale.getStringInLocale("time",this));
         quantityTV.setText(Locale.getStringInLocale("quantity",this));
+
+        totalDateTV.setText(Locale.getStringInLocale("date",this));
+        totalCowNameTV.setText(Locale.getStringInLocale("cow",this));
+        totalQuantityTV.setText(Locale.getStringInLocale("total", this));
+
         noDataWarning=Locale.getStringInLocale("no_data_received",this);
         times=Locale.getArrayInLocale("milking_times",this);
         if(times == null) {
@@ -115,6 +136,7 @@ public class MilkProcutionHistoryActivity extends SherlockActivity
         todayText=Locale.getStringInLocale("today",this);
         yesterdayText=Locale.getStringInLocale("yesterday",this);
         loadingPleaseWait=Locale.getStringInLocale("loading_please_wait",this);
+        backB.setText(Locale.getStringInLocale("back", this));
         initMenuText();
     }
 
@@ -139,6 +161,15 @@ public class MilkProcutionHistoryActivity extends SherlockActivity
         }
     }
 
+    @Override
+    public void onClick(View view) {
+        if(view == backB){
+            Intent intent = new Intent(this, MilkProductionActivity.class);
+            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
+            startActivity(intent);
+        }
+    }
+
     private class ProductionHistoryThread extends AsyncTask<String,Integer,String>
     {
         private ProgressDialog progressDialog;
@@ -199,6 +230,7 @@ public class MilkProcutionHistoryActivity extends SherlockActivity
                 {
                     JSONObject jsonObject=new JSONObject(result);
                     JSONArray historyArray=jsonObject.getJSONArray("history");
+                    addTotalTableRows(historyArray);
                     addHistoryTableRows(historyArray);
                 }
                 catch (JSONException e)
@@ -315,7 +347,7 @@ public class MilkProcutionHistoryActivity extends SherlockActivity
                 final View rowSeparator2=generateRowSeparator(3424+jsonObject.getInt("id")+3432,tableRowHeight);
                 tableRow.addView(rowSeparator2);
 
-                //TODO: check if name is null and use ear tag number
+                //check if name is null and use ear tag number
                 String nameText=jsonObject.getString("name");
                 if(nameText==null||nameText.equals(""))
                 {
@@ -355,4 +387,167 @@ public class MilkProcutionHistoryActivity extends SherlockActivity
             }
         }
     }
+
+    private void addTotalTableRows(JSONArray history){
+        //get totals for each cow
+
+        // array["cowName-ear_tag_number"]{"name","ear_tag_number","startDateMilliseconds","endDateMilliseconds","total"}
+        HashMap<String, String[]> cowMilkTotalArray = new HashMap<String, String[]>();
+
+        for(int i = 0; i < history.length(); i++){
+            try {
+                JSONObject jsonObject=history.getJSONObject(i);
+                String cowKey = jsonObject.getString("name")+"-"+jsonObject.getString("ear_tag_number");
+
+                //2. Initialize the cow data
+                //TODO: do conversions for other quantity types if they come up
+                    /*
+                    Since 1KG = 1L of milk you don't need to do any quantity conversions as of now
+                     */
+                String convertedQuantity = jsonObject.getString("quantity");
+
+                SimpleDateFormat dateFormat=new SimpleDateFormat("yyyy-MM-dd");
+                Date milkingDate=dateFormat.parse(jsonObject.getString("date"));
+                long milkingDateMilliseconds = milkingDate.getTime();
+
+                //1. check if the cow is in the cowMilkTotalArray
+                String[] currCowData = null;
+                if(cowMilkTotalArray.containsKey(cowKey)){
+                    currCowData = cowMilkTotalArray.get(cowKey);
+                }
+
+                if(currCowData == null){//no milk data from cow gotten until now in the loop
+                    /*
+                    Cow Data structure looks like:
+                        {"name","ear_tag_number","startDateMilliseconds","endDateMilliseconds","total", "quantity_type"}
+                     */
+
+                    Log.d(TAG, "Adding first time data for "+cowKey);
+                    String[] quantityTypesInEN = Locale.getArrayInLocale("quantity_types", this, Locale.LOCALE_ENGLISH);
+                    String[] quantityTypes = Locale.getArrayInLocale("quantity_types", this);
+                    //get translation of Litres in current locale
+                    String litresInLocale = null;
+                    for(int j = 0; j < quantityTypesInEN.length; j++){
+                        if(quantityTypesInEN[j].equals("Litres")){
+                            litresInLocale = quantityTypes[j];
+                        }
+                    }
+
+                    currCowData = new String[]{
+                            jsonObject.getString("name"),
+                            jsonObject.getString("ear_tag_number"),
+                            String.valueOf(milkingDateMilliseconds),
+                            String.valueOf(milkingDateMilliseconds),
+                            convertedQuantity,
+                            litresInLocale
+                    };
+                }
+                else{//Milk production from this cow already iterated through before i
+                    //3. Increment quantity and dates
+                    Log.d(TAG, "Updating data for "+cowKey);
+                    currCowData[4] = String.valueOf(Integer.parseInt(currCowData[4])+Integer.parseInt(convertedQuantity));
+
+                    //4. Check if current milk reading is in the date extremes (first date or last date)
+                    Log.d(TAG, "start time string = "+currCowData[2]);
+                    Log.d(TAG, "end time string = "+currCowData[3]);
+                    long startDateMilliseconds = Long.parseLong(currCowData[2]);
+                    long endDateMilliseconds = Long.parseLong(currCowData[3]);
+
+                    Log.d(TAG, "Current milking Date Milliseconds = "+String.valueOf(milkingDateMilliseconds));
+                    Log.d(TAG, "Start Date Milliseconds = "+String.valueOf(startDateMilliseconds));
+                    Log.d(TAG, "End Date Milliseconds = "+String.valueOf(endDateMilliseconds));
+
+                    if(milkingDateMilliseconds < startDateMilliseconds){
+                        startDateMilliseconds = milkingDateMilliseconds;
+                    }
+                    else if(milkingDateMilliseconds > endDateMilliseconds){
+                        endDateMilliseconds = milkingDateMilliseconds;
+                    }
+
+                    currCowData[2] = String.valueOf(startDateMilliseconds);
+                    currCowData[3] = String.valueOf(endDateMilliseconds);
+                }
+
+                cowMilkTotalArray.put(cowKey, currCowData);
+            }
+            catch (Exception e){
+                e.printStackTrace();
+            }
+        }
+
+        //render cow milk totals data to the user
+        int index = 0;
+        for(Map.Entry currCow : cowMilkTotalArray.entrySet()){
+
+            String[] currCowData = (String[]) currCow.getValue();
+
+            final TableRow tableRow=new TableRow(this);
+            tableRow.setId(1522+index);
+            int tableRowHeight=0;
+            int tableTextSideMargin=0;//4dp
+            int tableTextSize=0;//14dp
+            if(metrics.densityDpi==DisplayMetrics.DENSITY_XHIGH)
+            {
+                tableRowHeight=58;//initially 30
+                tableTextSideMargin=14;//initially 6
+                tableTextSize=16;//initially 21
+            }
+            else if(metrics.densityDpi==DisplayMetrics.DENSITY_HIGH)
+            {
+                tableRowHeight=44;//initially 30
+                tableTextSideMargin=14;//initially 6
+                tableTextSize=16;//initially 21
+            }
+            else if(metrics.densityDpi==DisplayMetrics.DENSITY_MEDIUM)
+            {
+                tableRowHeight=27;//initially 20
+                tableTextSideMargin=10;//initially 4
+                tableTextSize=15;//initially 14
+            }
+            else if(metrics.densityDpi==DisplayMetrics.DENSITY_LOW)
+            {
+                tableRowHeight=16;//initially 15
+                tableTextSideMargin=6;//initially 3
+                tableTextSize=15;//initially 11
+            }
+            else
+            {
+                tableRowHeight=58;//initially 30
+                tableTextSideMargin=14;//initially 6
+                tableTextSize=16;//initially 21
+            }
+            tableRow.setLayoutParams(new TableRow.LayoutParams(TableRow.LayoutParams.MATCH_PARENT,tableRowHeight));
+
+            Calendar startDate=new GregorianCalendar();
+            startDate.setTime(new Date(Long.valueOf(currCowData[2])));
+            Calendar endDate = new GregorianCalendar();
+            endDate.setTime(new Date(Long.valueOf(currCowData[3])));
+
+            String startDateText=String.valueOf(startDate.get(Calendar.DAY_OF_MONTH))+"/"+String.valueOf(startDate.get(Calendar.MONTH)+1)+"/"+startDate.get(Calendar.YEAR);
+            String endDateText=String.valueOf(endDate.get(Calendar.DAY_OF_MONTH))+"/"+String.valueOf(endDate.get(Calendar.MONTH) + 1)+"/"+endDate.get(Calendar.YEAR);
+
+            final TextView date=generateTextView(startDateText + " - " + endDateText, 3343 + index,tableRowHeight,tableTextSize);
+            tableRow.addView(date);
+
+            final View rowSeparator1=generateRowSeparator(index + 3322, tableRowHeight);
+            tableRow.addView(rowSeparator1);
+
+            String cowName = currCowData[0];
+            if(cowName == null || cowName.length() == 0){//if cow doesnt have a name, use its ear tag number
+                cowName = currCowData[1];
+            }
+            final TextView cow = generateTextView(cowName, 7343 + index,tableRowHeight,tableTextSize);
+            tableRow.addView(cow);
+
+            final View rowSeparator2=generateRowSeparator(index + 2322, tableRowHeight);
+            tableRow.addView(rowSeparator2);
+
+            final TextView quantity = generateTextView(currCowData[4]+" "+currCowData[5] ,922+index, tableRowHeight, tableTextSize);
+            tableRow.addView(quantity);
+
+            productionTotalTL.addView(tableRow, new TableLayout.LayoutParams(TableLayout.LayoutParams.MATCH_PARENT,tableRowHeight));
+
+            index++;
+        }
+    }
 }
diff --git a/MistroFarmer/src/main/res/layout/activity_milk_production_history.xml b/MistroFarmer/src/main/res/layout/activity_milk_production_history.xml
index 9db7fa0..ede52c6 100644
--- a/MistroFarmer/src/main/res/layout/activity_milk_production_history.xml
+++ b/MistroFarmer/src/main/res/layout/activity_milk_production_history.xml
@@ -8,62 +8,125 @@
 
     <ScrollView
         android:layout_height="match_parent"
-        android:layout_width="match_parent">
-        <TableLayout
-            android:id="@+id/production_history_tl"
-            android:layout_height="wrap_content"
+        android:layout_width="match_parent"
+        android:layout_weight="90">
+        <LinearLayout
             android:layout_width="match_parent"
-            android:stretchColumns="0,2,4,6">
-            <TableRow
-                android:layout_height="36dp"
-                android:layout_width="match_parent">
-                <TextView
-                    android:id="@+id/date_tv"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:textSize="@dimen/normal_text_size"
-                    android:gravity="center"/>
-                <View
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
+            <TableLayout
+                android:id="@+id/production_total_tl"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:stretchColumns="0,2,4">
+                <TableRow
                     android:layout_height="36dp"
-                    android:layout_width="1dp"
-                    android:background="@color/pressed_mistro"/>
-                <TextView
-                    android:id="@+id/time_tv"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:textSize="@dimen/normal_text_size"
-                    android:gravity="center"/>
-                <View
+                    android:layout_width="match_parent">
+                    <TextView
+                        android:id="@+id/total_date_tv"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:textSize="@dimen/normal_text_size"
+                        android:gravity="center"/>
+                    <View
                         android:layout_height="36dp"
                         android:layout_width="1dp"
                         android:background="@color/pressed_mistro"/>
-                <TextView
-                    android:id="@+id/cow_name_tv"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:textSize="@dimen/normal_text_size"
-                    android:gravity="center"/>
-                <View
+                    <TextView
+                        android:id="@+id/total_cow_name_tv"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:textSize="@dimen/normal_text_size"
+                        android:gravity="center"/>
+                    <View
                         android:layout_height="36dp"
                         android:layout_width="1dp"
                         android:background="@color/pressed_mistro"/>
-                <TextView
-                    android:id="@+id/quantity_tv"
-                    android:layout_height="wrap_content"
-                    android:layout_width="wrap_content"
-                    android:textSize="@dimen/normal_text_size"
-                    android:gravity="center"/>
-            </TableRow>
-            <TableRow
-                android:layout_height="4dp"
-                android:layout_width="match_parent">
-                <View
+                    <TextView
+                        android:id="@+id/total_quantity_tv"
+                        android:layout_height="wrap_content"
+                        android:layout_width="wrap_content"
+                        android:textSize="@dimen/normal_text_size"
+                        android:gravity="center"/>
+                </TableRow>
+                <TableRow
+                    android:layout_height="4dp"
+                    android:layout_width="match_parent">
+                    <View
+                        android:layout_height="1dp"
+                        android:layout_width="fill_parent"
+                        android:background="@color/pressed_mistro"
+                        android:layout_span="5"/>
+                </TableRow>
+            </TableLayout>
+            <View
+                android:layout_width="match_parent"
+                android:layout_height="1dp"
+                android:background="@color/pressed_mistro"
+                android:layout_marginTop="20dp"/>
+            <TableLayout
+                android:id="@+id/production_history_tl"
+                android:layout_height="wrap_content"
+                android:layout_width="match_parent"
+                android:stretchColumns="0,2,4,6">
+                <TableRow
+                    android:layout_height="36dp"
+                    android:layout_width="match_parent">
+                    <TextView
+                        android:id="@+id/date_tv"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:textSize="@dimen/normal_text_size"
+                        android:gravity="center"/>
+                    <View
+                        android:layout_height="36dp"
+                        android:layout_width="1dp"
+                        android:background="@color/pressed_mistro"/>
+                    <TextView
+                        android:id="@+id/time_tv"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:textSize="@dimen/normal_text_size"
+                        android:gravity="center"/>
+                    <View
+                        android:layout_height="36dp"
+                        android:layout_width="1dp"
+                        android:background="@color/pressed_mistro"/>
+                    <TextView
+                        android:id="@+id/cow_name_tv"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:textSize="@dimen/normal_text_size"
+                        android:gravity="center"/>
+                    <View
+                        android:layout_height="36dp"
+                        android:layout_width="1dp"
+                        android:background="@color/pressed_mistro"/>
+                    <TextView
+                        android:id="@+id/quantity_tv"
+                        android:layout_height="wrap_content"
+                        android:layout_width="wrap_content"
+                        android:textSize="@dimen/normal_text_size"
+                        android:gravity="center"/>
+                </TableRow>
+                <TableRow
+                    android:layout_height="4dp"
+                    android:layout_width="match_parent">
+                    <View
                         android:layout_height="1dp"
                         android:layout_width="fill_parent"
                         android:background="@color/pressed_mistro"
                         android:layout_span="7"/>
-            </TableRow>
-        </TableLayout>
-    </ScrollView>
+                </TableRow>
+            </TableLayout>
+        </LinearLayout>
 
+    </ScrollView>
+    <Button
+        android:id="@+id/back_b"
+        android:layout_width="@dimen/back_button_width"
+        android:layout_height="50dp"
+        android:paddingLeft="15dp"
+        android:paddingRight="15dp"
+        android:layout_weight="10"/>
 </LinearLayout>
diff --git a/MistroFarmer/src/main/res/values/strings_en.xml b/MistroFarmer/src/main/res/values/strings_en.xml
index 802b732..1b93116 100644
--- a/MistroFarmer/src/main/res/values/strings_en.xml
+++ b/MistroFarmer/src/main/res/values/strings_en.xml
@@ -231,4 +231,5 @@
     <string name="back_to_main_menu_en">Back to Main Menu</string>
     <string name="warning_en">Warning</string>
     <string name="action_will_be_canceled_en">If you choose to go back to the Main Menu the data you have entered in this screen will be lost. Press \'Okay\' if this is fine.</string>
+    <string name="total_en">Total</string>
 </resources>
diff --git a/MistroFarmer/src/main/res/values/strings_sw.xml b/MistroFarmer/src/main/res/values/strings_sw.xml
index 27ee87c..ae1cf71 100644
--- a/MistroFarmer/src/main/res/values/strings_sw.xml
+++ b/MistroFarmer/src/main/res/values/strings_sw.xml
@@ -231,4 +231,5 @@
     <string name="back_to_main_menu_sw">Rudi kwa Menu Kuu</string>
     <string name="warning_sw">Onyo</string>
     <string name="action_will_be_canceled_sw">Habari uliyoingiza katika ukurasa huu itapotea ukiamua kurudi kwa Menu Kuu. Bonyeza \'Sawa\' kama hii ni sawa na wewe.</string>
+    <string name="total_sw">Jumla</string>
 </resources>
-- 
2.7.1

